tests/testutils/aci-server/server.go

```diff
@@ -0,0 +1,306 @@
+	if auth == "" {
+		err := &httpError{
+			code:    http.StatusUnauthorized,
+			message: "No auth",
+		}
+		return "", err
+	}
+	parts := strings.Split(auth, " ")
+	if len(parts) != 2 {
+		err := &httpError{
+			code:    http.StatusBadRequest,
+			message: "Malformed auth",
+		}
+		return "", err
+	}
+	if parts[0] != authType {
+		err := &httpError{
+			code:    http.StatusUnauthorized,
+			message: "Wrong auth",
+		}
+		return "", err
+	}
+	return parts[1], nil
+}
+
+func (h *serverHandler) handleRequest(w http.ResponseWriter, r *http.Request) {
+	path := filepath.Base(r.URL.Path)
+	switch path {
+	case "/":
+		h.sendAcDiscovery(w)
+		h.sendMsg("  done.")
+	default:
+		if found := h.handleFile(w, path); found {
+			h.sendMsg("  done.")
+		}
+	}
+}
+
+func (h *serverHandler) sendAcDiscovery(w http.ResponseWriter) {
+	// TODO(krnowak): When appc spec gets the discovery over
+	// custom port feature, possibly take it into account here
+	indexHTML := `<meta name="ac-discovery" content="localhost https://localhost/{name}.{ext}">`
+	w.Write([]byte(indexHTML))
+}
+
+func (h *serverHandler) handleFile(w http.ResponseWriter, reqPath string) bool {
+	path, ok := h.fileSet[reqPath]
+	if !ok {
+		w.WriteHeader(http.StatusNotFound)
+		h.sendMsg("  not found.")
+		return false
+	}
+	if !h.canServe(reqPath, w) {
+		return false
+	}
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		h.sendMsg("  not found, but specified in fileset; bug?")
+		return false
+	}
+	w.Write(contents)
+	reqImagePath, isAsc := isPathAnImageKey(reqPath)
+	if isAsc {
+		delete(h.servedImages, reqImagePath)
+	} else {
+		h.servedImages[reqPath] = struct{}{}
+	}
+	return true
+}
+
+func (h *serverHandler) canServe(reqPath string, w http.ResponseWriter) bool {
+	if h.serverType != ServerQuay {
+		return true
+	}
+	reqImagePath, isAsc := isPathAnImageKey(reqPath)
+	if !isAsc {
+		return true
+	}
+	if _, imageAlreadyServed := h.servedImages[reqImagePath]; imageAlreadyServed {
+		return true
+	}
+	w.WriteHeader(http.StatusAccepted)
+	h.sendMsg("  asking to defer the download")
+	return false
+}
+
+func (h *serverHandler) sendMsg(msg string) {
+	select {
+	case h.msg <- msg:
+	default:
+	}
+}
+
+func isPathAnImageKey(path string) (string, bool) {
+	if strings.HasSuffix(path, ".asc") {
+		imagePath := strings.TrimSuffix(path, ".asc")
+		return imagePath, true
+	}
+	return "", false
+}
+
+type Server struct {
+	Msg     <-chan string
+	Conf    string
+	URL     string
+	handler *serverHandler
+	http    *httptest.Server
+}
+
+func (s *Server) Close() {
+	s.http.Close()
+	close(s.handler.msg)
+}
+
+func (s *Server) UpdateFileSet(fileSet map[string]string) {
+	s.handler.fileSet = fileSet
+}
+
+func NewServer(auth AuthType, msgCapacity int) *Server {
+	return NewServerFull(ServerQuay, auth, msgCapacity)
+}
+
+func NewQuayServer(auth AuthType, msgCapacity int) *Server {
+	return NewServerFull(ServerOrdinary, auth, msgCapacity)
```

Comment:
Nice catch. But git grepping shows that nothing uses the `NewQuayServer` function, and the only user of `NewServer` (the auth tests) actually does not care about it. I think I'll remove the `NewServer` and `NewQuayServer` functions, rename the `NewServerFull` function to `NewServer` and fix their uses accordingly.